pointmatcher/DataPointsFilters/Geometry.cpp

```diff
@@ -0,0 +1,170 @@
+    keepStructureness(Parametrizable::get<T>("keepStructureness"))
+{
+}
+
+// Compute
+template<typename T>
+typename PointMatcher<T>::DataPoints
+GeometryDataPointsFilter<T>::filter(
+	const DataPoints& input)
+{
+	DataPoints output(input);
+	inPlaceFilter(output);
+	return output;
+}
+
+// In-place filter
+template<typename T>
+void GeometryDataPointsFilter<T>::inPlaceFilter(
+	DataPoints& cloud)
+{
+	typedef typename DataPoints::View View;
+	typedef typename DataPoints::Label Label;
+	typedef typename DataPoints::Labels Labels;
+
+	const int pointsCount(cloud.features.cols());
+	const int descDim(cloud.descriptors.rows());
+	const unsigned int labelDim(cloud.descriptorLabels.size());
+
+	// Check that the required eigenValue descriptor exists in the pointcloud
+    if (!cloud.descriptorExists("eigValues"))
+    {
+        throw InvalidField("GeometryDataPointsFilter: Error, no eigValues found in descriptors.");
+    }
+
+    // Validate descriptors and labels
+	int insertDim(0);
+	for(unsigned int i = 0; i < labelDim ; ++i)
+		insertDim += cloud.descriptorLabels[i].span;
+	if (insertDim != descDim)
+		throw InvalidField("SurfaceNormalDataPointsFilter: Error, descriptor labels do not match descriptor data");
+
+	// Reserve memory for new descriptors
+	const int dimSphericality(1);
+	const int dimUnstructureness(1);
+	const int dimStructureness(1);
+
+	boost::optional<View> sphericality;
+	boost::optional<View> unstructureness;
+	boost::optional<View> structureness;
+
+	Labels cloudLabels;
+	cloudLabels.push_back(Label("sphericality", dimSphericality));
+	if (keepUnstructureness)
+		cloudLabels.push_back(Label("unstructureness", dimUnstructureness));
+	if (keepStructureness)
+		cloudLabels.push_back(Label("structureness", dimStructureness));
+
+    // Reserve memory
+	cloud.allocateDescriptors(cloudLabels);
+
+	// Get the views
+    const View eigValues = cloud.getDescriptorViewByName("eigValues");
+    if (eigValues.rows() != 3)  // And check the dimensions
+    {
+        throw InvalidField("GeometryDataPointsFilter: Error, the number of eigValues is not 3.");
+    }
+
+	sphericality = cloud.getDescriptorViewByName("sphericality");
+	if (keepUnstructureness)
+		unstructureness = cloud.getDescriptorViewByName("unstructureness");
+	if (keepStructureness)
+		structureness = cloud.getDescriptorViewByName("structureness");
+
+    // Iterate through the point cloud and evaluate the geometry
+	for (int i = 0; i < pointsCount; ++i)
+	{
+	    // look at the three eigen values
+        Vector eig_vals_col = eigValues.col(i);
+        // might be already sorted but sort anyway
+        std::sort(eig_vals_col.data(),eig_vals_col.data()+eig_vals_col.size());
+
+        // finally, evaluate the geometry
+        T sphericality_val;
+        T unstructureness_val;
+        T structureness_val;
+
+        //TODO: Is there a more suitable limit for considering the values almost-zero? (VK)
+        if (fabs(eig_vals_col(2)) < std::numeric_limits<T>::min() or
```

Comment:
You seem to be evaluating the smallest eigenvalues to determine the span of the points in space?

Can you write a comment explaining what you are doing here? If other developers want to extend your code, that will help them to get up to speed quickly.

Reply:
I've put there a reference to a paper